File: day12/digital_plumber.py
#!/usr/bin/env python3
from sys import argv
import networkx as nx
import re

# parse the input
G = nx.Graph()
filename = argv[1]
with open(filename) as f:
    for line in f:
        line = line.rstrip("\n")
        m = re.search('^(\d+) <-> (.*)$', line)
        fromv = int(m.group(1))
        tos = [int(x) for x in m.group(2).split(', ')]
        for tov in tos:
            G.add_edge(fromv, tov)

cc = list(nx.connected_components(G))
print("result1:", len(cc[0]))
print("result2:", len(cc))

File: day12/digital_plumber.pl
#!/usr/bin/env perl
use strict;
use warnings;
use feature qw(:5.24);
use experimental qw(signatures);
use Graph::Undirected;

my $g = Graph::Undirected->new;

# parse the input
while (<>) {
    chomp;
    $_ =~ /^(\d+) <-> (.*)$/;
    my $from = $1;
    my @tos = split /, /, $2;
    for my $to (@tos) {
        $g->add_edge($from, $to);
    }
}

my @cc = $g->connected_components();
for my $cc (@cc) {
    if (grep {$_ == 0} @$cc) {
        say "result1: ", scalar @$cc;
        last;
    }
}

say "result2: ", scalar @cc;
